Applications/Debugger/main.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2020, Itamar S. <itamar8910@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "DebugSession.h"
+#include <AK/Assertions.h>
+#include <AK/ByteBuffer.h>
+#include <AK/LogStream.h>
+#include <AK/StringBuilder.h>
+#include <AK/kmalloc.h>
+#include <LibC/sys/arch/i386/regs.h>
+#include <LibCore/File.h>
+#include <LibELF/ELFImage.h>
+#include <LibX86/Disassembler.h>
+#include <LibX86/Instruction.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+static int usage()
+{
+    printf("usage: sdb [command...]\n");
+    return 1;
+}
+
+OwnPtr<DebugSession> g_debug_session;
+
+static void handle_sigint(int)
+{
+    printf("Debugger: SIGINT\n");
+
+    // The destructor of DebugSession takes care of detaching
+    g_debug_session = nullptr;
+}
+
+VirtualAddress get_entry_point(int pid)
+{
+    auto path = String::format("/proc/%d/exe", pid);
+    dbg() << "path: " << path;
+    auto file = Core::File::construct(path);
+    if (!file->open(Core::File::ReadOnly)) {
+        fprintf(stderr, "Failed to open Debugged executable");
+        exit(1);
+    }
+    auto data = file->read_all();
+    dbg() << "data size:" << data.size();
+    ELFImage elf(data.data(), data.size());
```

Comment:
I broke this in #1750 but it should be easy enough to change to `#include <LibELF/Image>` and construct an `ELF:: Image` instead.

Side note for this part: might be faster to just mmap the file instead of opening it and reading it all into a heap buffer? I absolutely don't have the numbers to back that up though. Seems a waste to read the entire executable here and then throw it away...

Reply:
Alternatively, the start address should be available in the first page for sure. First 64 bytes even, as the elf header structure at the beginning of the file looks like this:
```c++
/* ELF Header */
typedef struct elfhdr {
    unsigned char e_ident[EI_NIDENT]; /* ELF Identification */
    Elf32_Half e_type;                /* object file type */
    Elf32_Half e_machine;             /* machine */
    Elf32_Word e_version;             /* object file version */
    Elf32_Addr e_entry; /** <--- here **/  /* virtual entry point */ 
    Elf32_Off e_phoff;                /* program header table offset */
...
} Elf32_Ehdr;
```
The debugger will need to be more clever about treating the virtual address in the elf header of the entry point as gospel in the future, because any ASLR or relocation that gets applied to the main executable will need to be added/accounted for.

The load address should be possible to calculate from /proc/[pid]/maps though?

Reply:
> Side note for this part: might be faster to just mmap the file instead of opening it and reading it all into a heap buffer? I absolutely don't have the numbers to back that up though. Seems a waste to read the entire executable here and then throw it away...

Absolutely! Looks like `MappedFile` should be great for this. 

> The load address should be possible to calculate from /proc/[pid]/maps though?

Yeah, should be possible. IIRC, gdb disables ASLR by default, so maybe we could just do that as well.